### DISCUSSION TITLE
problem about unwrap

### DISCUSSION THREAD

**Discussion by whlustb** [Elnaril/uniswap-universal-router-decoder#51] [Q&A] [answered]
I need to do swap from uni to eth, my code is as below, all is fine except the unwrap procedure, the weth continue to increase, but not unwrap to eth, and I have tried to change unwrap_weth 1st para from  FunctionRecipient.ROUTER, FunctionRecipient.SENDER, and 2nd para from 0 to int(0.03*10**18)) , but none work, it doesn't report any error, just not unwrap weth to eth, I have also tried use call only the unwrap function, not work as sell, any help for me to unwrap the weth to eth? many thank in advance.

**Comment by Elnaril** [comment]
Hi @whlustb ,

In order to receive ETH i/o WETH, I confirm that the first parameter should be `FunctionRecipient.SENDER`:

```python
.unwrap_weth(FunctionRecipient.SENDER, 0)
```

Apart from that, I don't see any obvious issue with your code.

**Comment by Elnaril** [comment]
Are you using this address for WETH on Sepolia: `0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14` ?

Also, maybe you could share the link to a transaction where you use `.unwrap_weth(FunctionRecipient.SENDER, 0)` ?

**Reply by Elnaril** [reply]
As far as I understand it, you would not get `int(0.01*10**18)` weth from your swap with `int(0.001 * 10**18)` uni ...

At block 5591582 (when your trx was done) we have:
```python
In [18]: quoter_contract.functions.quoteExactInput(path_1, 1000000000000000).call(block_identifier=5591582)
Out[18]: [5320970153792554, [183675120169770077943929389873], [1], 81425]

In [19]: 5320970153792554 / 10**18
Out[19]: 0.005320970153792554

```
So this line:
```python
.unwrap_weth(FunctionRecipient.SENDER, int(0.01*1018))
```
makes the transaction to revert.

You should really try:
```python
.unwrap_weth(FunctionRecipient.SENDER, 0)
```

**Comment by Elnaril** [comment, marked as answer]
1/ The `unwrap_weth()` method unwraps weth that are in the router's balance, not in you wallet.

2/ In your last transaction, you instructed the router to send the result of the swap directly to you, so there is no remaining weth to unwrap ...
The recipient of the swap method should be `FunctionRecipient.ROUTER` in this case.

**Reply by whlustb** [reply]
thank you so much, it works now, and I have paied for your coffe to your 0xe32A121C89121bF8B829D6d88BBEA89e5d88f002, thank u again.

**Reply by Elnaril** [reply]
Great!! :) And thanks a bunch for the coffees!! 👍
